// @cucumber/cucumber/lib/cli/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const helpers_1 = require("./helpers");
const install_validator_1 = require("./install_validator");
const configuration_builder_1 = require("./configuration_builder");
const run_1 = require("../run");
const argv_parser_1 = __importDefault(require("./argv_parser"));
class Cli {
    constructor({ argv, cwd, stdout, env, }) {
        this.argv = argv;
        this.cwd = cwd;
        this.stdout = stdout;
        this.env = env;
    }
    async run() {
        await (0, install_validator_1.validateInstall)(this.cwd);
        const fromArgv = argv_parser_1.default.parse(await (0, helpers_1.getExpandedArgv)({
            argv: this.argv,
            cwd: this.cwd,
        }));
        const configuration = await (0, configuration_builder_1.buildConfiguration)(fromArgv, this.env);
        const { success } = await (0, run_1.runCucumber)(configuration, {
            cwd: this.cwd,
            stdout: this.stdout,
            env: this.env,
        });
        return {
            shouldAdvertisePublish: !configuration.formats.publish &&
                !fromArgv.options.publishQuiet &&
                !(0, configuration_builder_1.isTruthyString)(this.env.CUCUMBER_PUBLISH_QUIET),
            shouldExitImmediately: fromArgv.options.exit,
            success,
        };
    }
}
exports.default = Cli;
