// cucumber-jest/dist/index.js
"use strict";

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js-stable/object/define-property");

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.default = void 0;

var _stringify = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/json/stringify"));

var _crypto = _interopRequireDefault(require("crypto"));

var _core = require("@babel/core");

var _babelPresetJest = _interopRequireDefault(require("babel-preset-jest"));

var _default = {
  canInstrument: false,
  getCacheKey: (fileData, filename, configString, {
    instrument
  }) => _crypto.default.createHash('md5').update('\0', 'utf8').update(fileData).update('\0', 'utf8').update(filename).update('\0', 'utf8').update(configString).update('\0', 'utf8').update('\0', 'utf8').update(instrument ? 'instrument' : '').digest('hex'),

  process(src, filePath, jestConfig) {
    const extensions = (0, _stringify.default)(jestConfig.moduleFileExtensions);
    const testFile = `
            const exec = require('cucumber-jest/dist/exec').exec;            
            exec('${jestConfig.cwd}', '${filePath}', ${extensions})
        `;
    const featureFile = (0, _core.transform)(testFile, {
      filename: filePath,
      presets: [_babelPresetJest.default],
      root: jestConfig.cwd
    });
    return featureFile ? featureFile.code : src;
  }

};
exports.default = _default;
module.exports = exports.default;
