// cucumber-jest/dist/parsers/table.js
"use strict";

var _Object$keys = require("@babel/runtime-corejs3/core-js-stable/object/keys");

var _Object$getOwnPropertySymbols = require("@babel/runtime-corejs3/core-js-stable/object/get-own-property-symbols");

var _filterInstanceProperty = require("@babel/runtime-corejs3/core-js-stable/instance/filter");

var _Object$getOwnPropertyDescriptor = require("@babel/runtime-corejs3/core-js-stable/object/get-own-property-descriptor");

var _forEachInstanceProperty = require("@babel/runtime-corejs3/core-js-stable/instance/for-each");

var _Object$getOwnPropertyDescriptors = require("@babel/runtime-corejs3/core-js-stable/object/get-own-property-descriptors");

var _Object$defineProperties = require("@babel/runtime-corejs3/core-js-stable/object/define-properties");

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js-stable/object/define-property");

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.createDataTable = createDataTable;
exports.generateExampleTableSteps = generateExampleTableSteps;
exports.parseExampleTable = parseExampleTable;
exports.parseVariables = parseVariables;

var _concat = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/concat"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime-corejs3/helpers/defineProperty"));

var _toConsumableArray2 = _interopRequireDefault(require("@babel/runtime-corejs3/helpers/toConsumableArray"));

var _asciiTable = _interopRequireDefault(require("ascii-table"));

var _space = require("../configs/space");

function ownKeys(object, enumerableOnly) { var keys = _Object$keys(object); if (_Object$getOwnPropertySymbols) { var symbols = _Object$getOwnPropertySymbols(object); if (enumerableOnly) symbols = _filterInstanceProperty(symbols).call(symbols, function (sym) { return _Object$getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { var _context6; _forEachInstanceProperty(_context6 = ownKeys(Object(source), true)).call(_context6, function (key) { (0, _defineProperty2.default)(target, key, source[key]); }); } else if (_Object$getOwnPropertyDescriptors) { _Object$defineProperties(target, _Object$getOwnPropertyDescriptors(source)); } else { var _context7; _forEachInstanceProperty(_context7 = ownKeys(Object(source))).call(_context7, function (key) { _Object$defineProperty(target, key, _Object$getOwnPropertyDescriptor(source, key)); }); } } return target; }

function createDataTable(rows) {
  const table = new _asciiTable.default();
  table.setHeading.apply(table, (0, _toConsumableArray2.default)(rows[0]));

  for (let i = 1; i < rows.length; i++) {
    table.addRow.apply(table, (0, _toConsumableArray2.default)(rows[i]));
  }

  const _iterable = table.toString().split('\n');

  let _result = [];

  for (let _key = 0, _length = _iterable.length, _value; _key < _length; ++_key) {
    _value = _iterable[_key];
    _result[_key] = _space.space + _value;
  }

  return _result.join('\n');
}

function generateExampleTableSteps(examples, scenario) {
  let _result2 = [];

  for (let _key2 = 0, _length2 = examples.length, _value2; _key2 < _length2; ++_key2) {
    var _context;

    _value2 = examples[_key2];
    const _iterable3 = scenario.steps;
    let _result3 = [];

    for (let _key3 = 0, _length3 = _iterable3.length, _value3; _key3 < _length3; ++_key3) {
      _value3 = _iterable3[_key3];
      _result3[_key3] = _objectSpread(_objectSpread(_objectSpread({}, _value3), _value3.docString ? {
        docString: _objectSpread(_objectSpread({}, _value3.docString), {}, {
          content: parseVariables(_value2, _value3.docString.content)
        })
      } : {}), {}, {
        text: parseVariables(_value2, _value3.text)
      });
    }

    _result2 = (0, _concat.default)(_context = []).call(_context, (0, _toConsumableArray2.default)(_result2), [_objectSpread(_objectSpread({}, scenario), {}, {
      name: parseVariables(_value2, scenario.name),
      steps: _result3
    })]);
  }

  return _result2;
}

function parseExampleTable(examples) {
  const _iterable4 = examples || [];

  let _result4 = [];

  for (let _key4 = 0, _length4 = _iterable4.length, _value4; _key4 < _length4; ++_key4) {
    var _context4;

    _value4 = _iterable4[_key4];
    const _iterable5 = _value4.tableHeader.cells;
    let _result5 = [];

    for (let _key5 = 0, _length5 = _iterable5.length, _value5; _key5 < _length5; ++_key5) {
      var _context2;

      _value5 = _iterable5[_key5];
      _result5 = (0, _concat.default)(_context2 = []).call(_context2, (0, _toConsumableArray2.default)(_result5), [_value5.value]);
    }

    const _keys = _result5;
    const _iterable6 = _value4.tableBody;
    let _result6 = [];

    for (let _key6 = 0, _length6 = _iterable6.length, _value6; _key6 < _length6; ++_key6) {
      var _context3;

      _value6 = _iterable6[_key6];
      _result6 = (0, _concat.default)(_context3 = []).call(_context3, (0, _toConsumableArray2.default)(_result6), [parseExampleTableKeyAndValues(_keys, _value6)]);
    }

    _result4 = (0, _concat.default)(_context4 = []).call(_context4, (0, _toConsumableArray2.default)(_result4), (0, _toConsumableArray2.default)(_result6));
  }

  return _result4;
}

function parseExampleTableKeyAndValues(keys, row) {
  let _result7 = [];

  for (let _key7 = 0, _length7 = keys.length, _value7; _key7 < _length7; ++_key7) {
    var _context5;

    _value7 = keys[_key7];
    _result7 = (0, _concat.default)(_context5 = []).call(_context5, (0, _toConsumableArray2.default)(_result7), [{
      key: _value7,
      value: row.cells[_key7].value
    }]);
  }

  return _result7;
}

function parseVariables(example, text) {
  let _result8 = text + '';

  for (let _key8 = 0, _length8 = example.length, _value8; _key8 < _length8; ++_key8) {
    _value8 = example[_key8];
    _result8 = _result8.replace(new RegExp(`<${_value8.key}>`), _value8.value);
  }

  return _result8;
}
